File: core/src/apps/ethereum/get_address.py
from typing import TYPE_CHECKING

from .keychain import PATTERNS_ADDRESS, with_keychain_from_path

if TYPE_CHECKING:
    from trezor.messages import EthereumAddress, EthereumGetAddress

    from apps.common.keychain import Keychain

    from .definitions import Definitions


@with_keychain_from_path(*PATTERNS_ADDRESS)
async def get_address(
    msg: EthereumGetAddress,
    keychain: Keychain,
    defs: Definitions,
) -> EthereumAddress:
    from trezor import TR
    from trezor.messages import EthereumAddress
    from trezor.ui.layouts import show_address

    from apps.common import paths
    from apps.common.address_mac import get_address_mac

    from .helpers import address_from_bytes

    address_n = msg.address_n  # local_cache_attribute

    await paths.validate_path(keychain, address_n)

    node = keychain.derive(address_n)

    address = address_from_bytes(node.ethereum_pubkeyhash(), defs.network)

    slip44_id = address_n[1]  # it depends on the network (ETH vs ETC...)
    mac = get_address_mac(address, paths.unharden(slip44_id), address_n, keychain)

    if msg.show_display:
        coin = "ETH"
        await show_address(
            address,
            subtitle=TR.address__coin_address_template.format(coin),
            path=paths.address_n_to_str(address_n),
            account=paths.get_account_name(
                coin, address_n, PATTERNS_ADDRESS, slip44_id
            ),
            chunkify=bool(msg.chunkify),
        )

    return EthereumAddress(address=address, mac=mac)

File: sdk/apps/ethereum/src/get_address.rs
use crate::{
    common::{COIN, get_eth_pubkey_hash},
    definitions::Definitions,
    helpers::address_from_bytes,
    layout::addr_pad,
    paths::{Bip32Path, PATTERNS_ADDRESS},
    proto::{
        common::button_request::ButtonRequestType,
        ethereum::{Address, GetAddress},
    },
    uformat,
};
use trezor_app_sdk::{Result, ResultExt, crypto, ui};

/// Ethereum uses Bitcoin xpub format
pub(crate) fn get_address(msg: GetAddress) -> Result<Address> {
    let dp: Bip32Path = Bip32Path::from_slice(&msg.address_n);

    let slip44 = dp.slip44();
    let encoded_network = msg.encoded_network.as_deref();
    let definitions = Definitions::from_encoded(encoded_network, None, None, slip44)
        .context("Failed to decode definitions")?;

    crypto::verify_derivation_path(dp.as_slice(), encoded_network, None, None)
        .context("Failed to verify derivation path")?;

    let pubkey_hash = get_eth_pubkey_hash(&dp).context("Failed to get eth pubkey hash")?;
    let address = address_from_bytes(&pubkey_hash, Some(definitions.network()))
        .context("Failed to convert bytes to address")?;
    let slip44_id = &msg.address_n[1]; // it depends on the network (ETH vs ETC...)
    let mac = crypto::get_address_mac(dp.as_ref(), &address, encoded_network)
        .context("Failed to get address MAC")?;

    if let Some(true) = msg.show_display {
        // TODO: template translation
        let subtitle = uformat!("{} address", COIN);
        let account_name = dp
            .get_account_name(COIN, &PATTERNS_ADDRESS, *slip44_id)
            .ok_or(crate::Error::DataError("Failed to get account name"))?;
        let chunkify = msg.chunkify.unwrap_or_default();
        ui::error_if_not_confirmed(ui::show_address(ui::ShowAddress::new(
            &addr_pad(&address, chunkify)?,
            &address,
            None,
            Some(subtitle.as_str()),
            Some(account_name.as_str()),
            Some(&dp.format_path()),
            &[],
            chunkify,
            ButtonRequestType::Other.into(),
        ))?)
        .context("Address was not confirmed")?;

        ui::show_success(ui::ShowSuccess::new(
            tr!("words__title_done"),
            tr!("address__confirmed"),
            tr!("instructions__continue_in_app"),
            Some(3200),
            None,
            ButtonRequestType::Other.into(),
        ))?;
    }

    let res = Address {
        address: Some(address),
        mac: Some(mac.to_vec()),
        ..Default::default()
    };

    Ok(res)
}
